GitHub ALM & GitHub Binding

ViktorLindgren95 opened the pull request:
( I notice that the stuff from my other PR is in here aswell)

### Added two functionalities

#### GitHub ALM
Added the ability to add a Github ALM / Devops Platform Integration to your sonarqube instance
This comes with both documentation and testing which, including a live integration test on my local environment
Importing isnt supported right now, I will hopefully add this over the comming days if i have the spare time

#### GitHub Binding
Added the ability to add a relation between your Sonarqube project and your GitHub repository
This comes with both documentation and testing which, including a live integration test on my local environment
Importing isnt supported right now, I will hopefully add this over the comming days if i have the spare time


## Breaking change
ALM settings require sonarqube enterprise version, therefor i have changed the testing image to enterprise. It still works to do tests like we have been doing with an enterprise image aswell

ViktorLindgren95 commented:
Oh ok , i thought it was specifically enterprise . My bad 😊

On Thu, Feb 2, 2023, 18:03 jwilliams-fn ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In Makefile
> <https://github.com/jdamata/terraform-provider-sonarqube/pull/119#discussion_r1094818244>
> :
>
> > @@ -1,7 +1,7 @@
>  export GO111MODULE=on
>  export TF_LOG=DEBUG
>  SRC=$(shell find . -name '*.go')
> -SONARQUBE_IMAGE?=sonarqube:latest
> +SONARQUBE_IMAGE?=sonarqube:enterprise
>
> alm_settings APIs only require the developer or greater sonarqube.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jdamata/terraform-provider-sonarqube/pull/119#pullrequestreview-1281484757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFRJY4C73CMVH3Z5XVCK67TWVPSEZANCNFSM6AAAAAAUO7MQCE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>

ViktorLindgren95 commented on Makefile:
```diff
@@ -1,7 +1,7 @@
 export TF_LOG=DEBUG
 SRC=$(shell find . -name '*.go')
-SONARQUBE_IMAGE?=sonarqube:latest
+SONARQUBE_IMAGE?=sonarqube:enterprise
```
Resolved

jdamata commented on sonarqube/resource_sonarqube_project_main_branch.go:
```diff
@@ -0,0 +1,139 @@
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/project_branches/rename"
+	sonarQubeURL.RawQuery = url.Values{
+		"name":    []string{"main"},
```
Please get the default branch name from `[]string{d.Get("name").(string)}` instead of hardcoding main,

ViktorLindgren95 replied:
The idea of delete was to revert the branch back to what is default for the sonarqube instance. That is main unless the setting has been changed by an administrator. Perhaps we can fetch it via the api for settings because we know the setting name. But i dont know how that would work as i dont write too much terraform provider. if that makes sense

jdamata replied:
ah ok, i misunderstood. Does sonarqubes default branch change depending on which git provider is being used? I think if the default is always main we can leave this as a TODO. Shouldn't be too much work to implement.

If its possible to fetch this config via https://next.sonarqube.com/sonarqube/web_api/api/settings/values then that would be awesome. In the provider you would just make another http call on deletion to fetch that config.

This example is for on creation but its a similar idea. https://github.com/jdamata/terraform-provider-sonarqube/pull/120/files#diff-5ad43f98cfd73869c45ce8eede0af3226ba13143149e948a519dc6c363ed2f28R42 . First we call the isLocal() function to get some information about a user account before creating it.

ViktorLindgren95 replied:
> ah ok, i misunderstood. Does sonarqubes default branch change depending on which git provider is being used? I think if the default is always main we can leave this as a TODO. Shouldn't be too much work to implement.
> 

![image](https://user-images.githubusercontent.com/23239792/216664223-b9267296-f550-4753-891e-9aaad98fc914.png)
Sonarqube has a Rule that sets the default main branch name , this is by default main but it can be changed by the user. Be best solution would be if we can , when deleting the github binding that we just revert to the value of this setting. I hardcoded it main because i dont know how that logic would look like, or if its ok to call an api call that is actually unrelated to the deletetion itself if that makes sense. But because we already have functions in place to read rules from the settings provider then we could reuse that logic to fetch the rule. If you write a pseudo perhaps i can implement it, it shouldnt be too hard i think? But i think we can leave it as todo for now, i dont mind comming back to this functionality, as im probably also doing imports in a future pr. I just need this for work right now ^^ 


> If its possible to fetch this config via https://next.sonarqube.com/sonarqube/web_api/api/settings/values then that would be awesome. In the provider you would just make another http call on deletion to fetch that config.
> 

Its possible to fetch it as i state above. We could fetch it as you talk about below in like an init step perhaps so we only have to call it once if thats possible.

> This example is for on creation but its a similar idea. https://github.com/jdamata/terraform-provider-sonarqube/pull/120/files#diff-5ad43f98cfd73869c45ce8eede0af3226ba13143149e948a519dc6c363ed2f28R42 . First we call the isLocal() function to get some information about a user account before creating it.

I dont want to have to make the user provide me with the branch name they wanna revert back to. If you were talking about specifically askign the user to provide like a "default_branch_name" attribute field we can read

ViktorLindgren95 replied:
If you think my logic is wrong do tell me, This is the first time im writing anything for terraform and i've just recently learned using it. Thats why i havent done the import functions yet because i havent used them peronsally yet in my projects , so i dont actually know fully what it does 😄

jdamata replied:
> I hardcoded it main because i dont know how that logic would look like, or if its ok to call an api call that is actually unrelated to the deletetion itself if that makes sense.

Yep definitely ok to make the http call to fetch the setting. It is related to deletion since we need that data. 

Im good leaving it as a TODO for now

ViktorLindgren95 replied:
Then lets agree on leaving it a TODO for now. I will revisit it when i have more time :)

jdamata commented:
tests are failing because we also need to update github pipelines to use sonarqube:developer image. Can you also do this please :)

ViktorLindgren95 commented:
> tests are failing because we also need to update github pipelines to use sonarqube:developer image. Can you also do this please :)

Updated the both tests. I also changed the tag specifically for the developer one to LTS dynamic, I Can change it back to specific 8.9 if you so desire, they are the same for now

ViktorLindgren95 commented on sonarqube/resource_sonarqube_alm_github.go:
```diff
@@ -0,0 +1,188 @@
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"appid": {
```
@jdamata Im aware of these attributes not following the style guide for terraform ( which i believe is snake case) and i will solve this tommorow after i got some friday night sleep 😄

jdamata replied:
sounds good :)

ViktorLindgren95 replied:
@jdamata Its done ,i also added imports. Do note that the tests are bound to 8.9 now, While LTS is 9.8 i think because sonarqube changed it yesterday

ViktorLindgren95 commented:
Local testing has hit a snag because sonarqube changed LTS to 9.9 over night

ViktorLindgren95 commented:
@jdamata Ready for review 👍

jdamata commented:
Thank you for the contribution. Ill cut a new release in a few minutes so folks can start using the new resources
